// web/app/api/ship/route.ts
import { NextResponse } from "next/server";
import { PrismaClient } from "@prisma/client";
import { Octokit } from "@octokit/rest";

const prisma = new PrismaClient();

export async function POST(req: Request) {
  const { runId, owner, repo } = await req.json();

  const run = await prisma.run.findUnique({
    where: { id: runId },
    include: { agent: true, scenarios: true },
  });

  if (!run) {
    return NextResponse.json({ error: "Run not found" }, { status: 404 });
  }

  const octokit = new Octokit({ auth: process.env.GITHUB_TOKEN });
  const branchName = `forge/${run.agent.name}-${runId.slice(0, 8)}`;

  // Build PR body with sim report
  const passed = run.scenarios.filter((s) => s.verdict === "pass").length;
  const failed = run.scenarios.filter((s) => s.verdict === "fail").length;
  const prBody = `## Forge Simulation Report

**Agent:** ${run.agent.name}
**Sigma Score:** σ = ${run.sigmaScore?.toFixed(2) ?? "N/A"}
**Scenarios:** ${passed} passed, ${failed} failed out of ${run.scenarios.length}

### Scenario Results

| # | Input | Verdict | Score |
|---|-------|---------|-------|
${run.scenarios
  .map(
    (s, i) =>
      `| ${i + 1} | ${s.input.slice(0, 60)}... | ${s.verdict === "pass" ? "✅ pass" : "❌ fail"} | ${s.score.toFixed(2)} |`
  )
  .join("\n")}

*Generated by [Forge](https://github.com/your-org/forge)*`;

  try {
    // Get default branch SHA
    const { data: repoData } = await octokit.repos.get({ owner, repo });
    const defaultBranch = repoData.default_branch;
    const { data: ref } = await octokit.git.getRef({
      owner,
      repo,
      ref: `heads/${defaultBranch}`,
    });

    // Create branch
    await octokit.git.createRef({
      owner,
      repo,
      ref: `refs/heads/${branchName}`,
      sha: ref.object.sha,
    });

    // Open PR
    const { data: pr } = await octokit.pulls.create({
      owner,
      repo,
      title: `feat: ${run.agent.name} agent — σ ${run.sigmaScore?.toFixed(1) ?? "?"}`,
      head: branchName,
      base: defaultBranch,
      body: prBody,
    });

    return NextResponse.json({ prUrl: pr.html_url, prNumber: pr.number });
  } catch (err) {
    return NextResponse.json({ error: String(err) }, { status: 500 });
  }
}
